tests/data/iterators/lazy_iterator_test.py

```diff
@@ -0,0 +1,79 @@
+from allennlp.data.token_indexers import SingleIdTokenIndexer
+
+class LazyIteratorTestCase(AllenNlpTestCase):
+    def setUp(self):
+        super(LazyIteratorTestCase, self).setUp()
+        self.token_indexers = {"tokens": SingleIdTokenIndexer()}
+        self.vocab = Vocabulary()
+        self.this_index = self.vocab.add_token_to_namespace('this')
+        self.is_index = self.vocab.add_token_to_namespace('is')
+        self.a_index = self.vocab.add_token_to_namespace('a')
+        self.sentence_index = self.vocab.add_token_to_namespace('sentence')
+        self.another_index = self.vocab.add_token_to_namespace('another')
+        self.yet_index = self.vocab.add_token_to_namespace('yet')
+        self.very_index = self.vocab.add_token_to_namespace('very')
+        self.long_index = self.vocab.add_token_to_namespace('long')
+        self.instances = [
+                self.create_instance(["this", "is", "a", "sentence"]),
+                self.create_instance(["this", "is", "another", "sentence"]),
+                self.create_instance(["yet", "another", "sentence"]),
+                self.create_instance(["this", "is", "a", "very", "very", "very", "very", "long", "sentence"]),
+                self.create_instance(["sentence"]),
+                ]
+        self.dataset = LazyDataset(lambda: (i for i in self.instances))
+        self.dataset.index_instances(self.vocab)
+
+    def create_instance(self, str_tokens: List[str]):
+        tokens = [Token(t) for t in str_tokens]
+        instance = Instance({'text': TextField(tokens, self.token_indexers)})
+        instance.index_fields(self.vocab)
+        return instance
+
+    def assert_instances_are_correct(self, candidate_instances):
+        # First we need to remove padding tokens from the candidates.
+        # pylint: disable=protected-access
+        candidate_instances = [tuple(w for w in instance if w != 0) for instance in candidate_instances]
+        expected_instances = [tuple(instance.fields["text"]._indexed_tokens["tokens"])
+                              for instance in self.instances]
+        assert set(candidate_instances) == set(expected_instances)
+
+
+class TestLazyIterator(LazyIteratorTestCase):
+    # We also test some of the stuff in `DataIterator` here.
+    def test_yield_one_epoch_iterates_over_the_data_once(self):
+        iterator = LazyIterator(batch_size=2)
+        batches = list(iterator(self.dataset, num_epochs=1))
+        # We just want to get the single-token array for the text field in the instance.
+        instances = [tuple(instance.data.cpu().numpy())
+                     for batch in batches
+                     for instance in batch['text']["tokens"]]
+        assert len(instances) == 5
+        self.assert_instances_are_correct(instances)
+
+    def test_call_iterates_over_data_forever(self):
+        generator = LazyIterator(batch_size=2)(self.dataset)
+        batches = [next(generator) for _ in range(18)]  # going over the data 6 times
```

Comment:
I'm trying to figure out where in the code you get the infinite looping behavior, but I can't find it.  What am I missing?

Reply:
in the base class, `__call__` just does

```
while True:
  yield from self._yield_one_epoch(...)
```

and each call to `LazyIterator._yield_one_epoch()` calls `self.generator()` to get a fresh iterator over the whole dataset.

Reply:
Ah, great, thanks.  So, we need to be able to handle cases where one "epoch" does not exhaust the generator, and the next epoch just keeps going from where the last epoch left off.  This is pretty important, and related to my [most recent comment above](https://github.com/allenai/allennlp/pull/660#discussion_r159767396).  The iterator should have a `num_instances_per_epoch` parameter, or something (maybe `num_batches_per_epoch` instead...), and that's used here.  And then when the generator is exhausted, you start it again.  I think there are a few ways to get this kind of behavior, and I'm not sure what the best one is.